earthsim/__init__.py

```diff
@@ -0,0 +1,45 @@
+import param
+
+def params_from_kwargs(**kwargs):
+    """
+    Utility to promote keywords with literal values to the appropriate
+    parameter type with the specified default value unless the value is
+    already a parameter.
+    """
+    params = {}
+    for k, v in kwargs.items():
+        kws = dict(default=v)
+        if isinstance(v, param.Parameter):
+            params[k] = v
+        elif isinstance(v, bool):
+            params[k] = param.Boolean(**kws)
+        elif isinstance(v, int):
+            params[k] = param.Integer(**kws)
+        elif isinstance(v, float):
+            params[k] = param.Number(**kws)
+        elif isinstance(v, str):
+            params[k] = param.String(**kws)
+        elif isinstance(v, dict):
+            params[k] = param.Dict(**kws)
+        elif isinstance(v, tuple):
+            params[k] = param.Tuple(**kws)
+        elif isinstance(v, list):
+            params[k] = param.List(**kws)
+        elif isinstance(v, np.ndarray):
+            params[k] = param.Array(**kws)
+        else:
+            params[k] = param.Parameter(**kws)
+    return params
+
+
+def global_params(**kwargs):
+    """
+    Utility to easily define a parameterized class with a chosen set of
+    parameters specified as keyword arguments. The resulting object can
+    be used to parameterize a notebook, display corresponding widgets
+    with parambokeh and control the workflow from the command line.
+    """
+    name = 'GlobalParams'
```

Comment:
What makes this "global_params"?  It doesn't seem to have anything global about what it does or how it works.  How about just calling it "parameterized()" with a name "parameterized", since it simply returns a Parameterized class?  I strongly prefer objects to be named with an accurate description of what they do, to make it clear that no special magic is happening.

Reply:
(Yes, I do know that it's being *used* as a set of global parameters, later, but that shouldn't affect how it's named *here*.  The globality can be conveyed by a local variable name, if needed, at the point of use.)

Reply:
The name ``global_params`` was taken from topographica as the way it is *used* is very similar in concept. I agree that what it *is* is nothing global or special though 1) this doesn't define methods so the result is just a collection of parameters and nothing more 2) it is intended to be a singleton class for use at the top of a script/notebook to parameterize the rest of the workflow. What other uses do we have for classes that are just collections of parameters without code?

I don't mind renaming it ``parameterized`` though 1) that might get confused with ``param.Parameterized`` 2) the handle given by the user should still reflect the intent e.g:

```python
nbparams = parameterized(...)
```

Reply:
Out of those two options, I think ``parameters`` is definitely the better choice. I'll update to using that name now.

Reply:
BTW, we're not actually using the fact that it's a singleton right now; one could imagine having some counter and increasing the name by it each time, if we wanted to have more than one such object allowed.  So I don't think the singleton aspect should be part of the name.

Reply:
It is used as a singleton in once specific way: if you partition your parameter sets into two classes and try to set them separately via ``JSONInit``in two different calls to ``parambokeh.Widgets``, then you will see warnings as all the parameters are supplied at once at the command line (in a single environment variable). Each JSONInit call should be issuing warnings about the parameters that it cannot set that have been specified on the command line. For this reason, if you want to avoid these warnings, it is best to treat it as a singleton.

Reply:
Seems like that's true whether it's a singleton or not, or whether you use multiple explicit Parameterized classes -- unless the parameters supplied are present in all such objects, there will be an error.  What makes it a single object is that we're applying a single set of parameters from the envt to it, which again is about how it is used, not about how it's defined.

I was about to say that we should supply a way for the method to avoid being a singleton by letting the user provide a name, but now I see there's a problem if the user does that:

```
name = 'Parameters'
params = params_from_kwargs(**kwargs)
params['name'] = param.String(default=name)
```

I.e., if the user passes in a 'name' kw, it will silently be overwritten.  That's dangerous, as it's quite likely that eventually a user will have a parameter named 'name'.   Seems like we could instead check for the name in the supplied kwargs and use that, using 'Parameters' only if not found, which would serve a double purpose of providing a way to have arbitrarily many different classes created in this way, and also avoids clobbering a user-defined `name`.

Reply:
> ... unless the parameters supplied are present in all such objects, there will be an error.

That isn't necessarily true: you can target ``JSONInit`` to [specifically named classes](https://github.com/ioam/paramnb/blob/master/doc/JSONInit.ipynb) (Example 2 onwards), in which case you could partition the objects. Currently the notebooks in this PR are using the simplest approach in Example 1. I *could* have tried to fix the class name for parameters using the other more complex formats but I didn't which means I can use name as you suggest.

Reply:
Addressed in 9e38275

Reply:
Looks good, thanks.